File: tests/test_database_migration.py
#!/usr/bin/env python3
"""Test database schema, FTS5 full-text index, and migration integrity."""
import os
import sqlite3
import subprocess
import sys
import unittest
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))
import db

DB_PATH = db.DB_PATH
MIGRATE_SCRIPT = os.path.expanduser("~/.konoha/migrate.py")

class TestDatabaseMigration(unittest.TestCase):
    def setUp(self):
        # Database must exist
        self.assertTrue(os.path.exists(DB_PATH), f"Database not found at {DB_PATH}. Run migration first.")

    def test_schema_tables(self):
        """Verify that all core tables exist in the SQLite database."""
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        # Get all tables
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
        tables = {row[0] for row in cursor.fetchall()}
        conn.close()
        
        expected_tables = {"skills", "tool_calls", "agents", "bridges", "skill_chunks"}
        for t in expected_tables:
            self.assertIn(t, tables, f"Table '{t}' is missing from schema.")

    def test_fts5_indexing(self):
        """Verify that the FTS5 virtual table matches the search pattern."""
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        # Test basic match query
        try:
            cursor.execute("SELECT name, content FROM skills_fts WHERE content MATCH 'security' LIMIT 1;")
            result = cursor.fetchone()
            if result:
                self.assertIsNotNone(result[0])
                self.assertIsNotNone(result[1])
        except sqlite3.OperationalError as e:
            self.fail(f"FTS5 MATCH query failed: {e}")
        finally:
            conn.close()

    def test_migration_execution(self):
        """Verify that the production migration script executes cleanly in isolation."""
        import tempfile
        with tempfile.TemporaryDirectory() as tmp:
            temp_db = os.path.join(tmp, "konoha-test-migrate.db")
            skills_dir = os.path.join(tmp, "skills", "genin-skill")
            os.makedirs(skills_dir)
            with open(os.path.join(skills_dir, "SKILL.md"), "w", encoding="utf-8") as f:
                f.write("---\nname: genin-skill\ndescription: exploration\n---\n# Scout\n")
            proc = subprocess.run(
                [sys.executable, MIGRATE_SCRIPT, "--db-path", temp_db, "--skills-dir", os.path.dirname(skills_dir), "--require-skill", "genin-skill"],
                capture_output=True,
                text=True,
                timeout=25
            )
            self.assertEqual(proc.returncode, 0, f"Migration script execution failed: {proc.stderr}\nStdout: {proc.stdout}")

if __name__ == "__main__":
    unittest.main()

File: tests/test_database_migration.js
#!/usr/bin/env node
'use strict';

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { spawnSync } = require('child_process');
const Database = require('better-sqlite3');
const { DB_PATH } = require('../src/db');

const MIGRATE_SCRIPT = fs.existsSync(path.join(os.homedir(), '.konoha', 'migrate.js'))
  ? path.join(os.homedir(), '.konoha', 'migrate.js')
  : path.join(__dirname, '..', 'src', 'migrate.js');

function testSchemaTables() {
  assert.ok(fs.existsSync(DB_PATH), `Database not found at ${DB_PATH}. Run migration first.`);
  const db = new Database(DB_PATH, { readonly: true });
  const rows = db.prepare("SELECT name FROM sqlite_master WHERE type='table';").all();
  const tables = new Set(rows.map(r => r.name));
  db.close();

  const expectedTables = ['skills', 'tool_calls', 'agents', 'bridges', 'skill_chunks'];
  for (const t of expectedTables) {
    assert.ok(tables.has(t), `Table '${t}' is missing from schema.`);
  }
}

function testFts5Indexing() {
  assert.ok(fs.existsSync(DB_PATH), `Database not found at ${DB_PATH}.`);
  const db = new Database(DB_PATH, { readonly: true });
  try {
    const result = db.prepare("SELECT name, content FROM skills_fts WHERE content MATCH 'security' LIMIT 1;").get();
    if (result) {
      assert.ok(result.name !== undefined);
      assert.ok(result.content !== undefined);
    }
  } finally {
    db.close();
  }
}

function testMigrationExecution() {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'konoha-test-migrate-'));
  try {
    const tempDb = path.join(tmp, 'konoha-test-migrate.db');
    const skillsDir = path.join(tmp, 'skills', 'genin-skill');
    fs.mkdirSync(skillsDir, { recursive: true });
    fs.writeFileSync(
      path.join(skillsDir, 'SKILL.md'),
      '---\nname: genin-skill\ndescription: exploration\n---\n# Scout\n',
      'utf-8'
    );

    const proc = spawnSync(
      process.execPath,
      [MIGRATE_SCRIPT, '--db-path', tempDb, '--skills-dir', path.dirname(skillsDir), '--require-skill', 'genin-skill', '--skip-embeddings'],
      { encoding: 'utf-8', timeout: 60000 }
    );
    assert.strictEqual(proc.status, 0, `Migration script execution failed: ${proc.stderr}\nStdout: ${proc.stdout}`);
  } finally {
    fs.rmSync(tmp, { recursive: true, force: true });
  }
}

console.log('Running test_database_migration.js...');
testSchemaTables();
testFts5Indexing();
testMigrationExecution();
console.log('test_database_migration.js passed cleanly.');
